controllers/component_build_controller.go

```diff
@@ -126,34 +145,176 @@ func (r *ComponentBuildReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		return ctrl.Result{}, nil
 	}
 
-	// Set initial build annotation to prevent next builds
-	component.Annotations[InitialBuildAnnotationName] = "true"
-	if err := r.Client.Update(ctx, &component); err != nil {
-		return ctrl.Result{}, err
-	}
+	if val, ok := component.Annotations[gitops.PaCAnnotation]; ok && val == "1" {
+		r.Log.Info("PaC enabled")
+		if err := r.EnsurePACRepository(ctx, component); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.GeneratePullRequest(ctx, component); err != nil {
+			return ctrl.Result{}, err
+		}
 
-	if err := r.SubmitNewBuild(ctx, component); err != nil {
-		// Try to revert the annotation
-		if err := r.Client.Get(ctx, req.NamespacedName, &component); err == nil {
-			component.Annotations[InitialBuildAnnotationName] = "false"
-			if err := r.Client.Update(ctx, &component); err != nil {
+		// Set initial build annotation to prevent next builds
+		component.Annotations[InitialBuildAnnotationName] = "true"
+		if err := r.Client.Update(ctx, &component); err != nil {
+			return ctrl.Result{}, err
+		}
+	} else {
+		// Set initial build annotation to prevent next builds
+		component.Annotations[InitialBuildAnnotationName] = "true"
+		if err := r.Client.Update(ctx, &component); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if err := r.SubmitNewBuild(ctx, component); err != nil {
+			// Try to revert the annotation
```

Comment:
so does this mean that if we can't build via pipeline as code, we try to build "normally" ?

If so, I think some logging and generating some events in the user's namesapce / workspace would be good here.

Reply:
No, we don't. As we are in switching to PaC phase, we currently support both builds: trigger (old) and PaC (new). The code you asked about is completely related to the trigger build. We use an annotation to mark a component that initial build happened (and some other preparations done). But if something in the preparation or build submition failed, we just delete (revert) the annotation, so next reconcile loop will retry. The annotation is added before all the preparations and build submition because there is a risk that the component is changed and the annotation update will fail, so the initial build would be run twice. To avoid that, we set annotation first and in case of a preparation error (a rare case) just revert it.